File: src/test/java/io/protobase/f7/formulas/info/ISTEXTTest.java
package io.protobase.f7.formulas.info;

import io.protobase.f7.errors.NAException;
import io.protobase.f7.models.Grid;
import io.protobase.f7.testutils.TestFormula;
import org.junit.Test;

import static com.google.common.truth.Truth.assertThat;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

public class ISTEXTTest extends TestFormula {
  @Test
  public void test_number() {
    assertThat(ISTEXT.SELF.apply(null, 10.0)).isEqualTo(false);
  }

  @Test
  public void test_string() {
    assertThat(ISTEXT.SELF.apply(null, "String")).isEqualTo(true);
  }

  @Test
  public void test_boolean() {
    assertThat(ISTEXT.SELF.apply(null, true)).isEqualTo(false);
    assertThat(ISTEXT.SELF.apply(null, false)).isEqualTo(false);
  }

  @Test
  public void test_error() {
    assertThat(ISTEXT.SELF.apply(null, new NAException())).isEqualTo(false);
  }

  @Test
  public void test_lookup() {
    ISTEXT F = new ISTEXT(lookup, collateralLookup);
    when(collateralLookup.apply(A1, M22_RANGE)).thenReturn("Nope");
    assertThat(F.apply(A1, M22_RANGE)).isEqualTo(true);
    verify(collateralLookup).apply(A1, M22_RANGE);
    verifyNoMoreInteractions(collateralLookup);
  }

  @Test
  public void test_grid() {
    assertThat(ISTEXT.SELF.apply(null, Grid.builder().add(0, 0, "Yes").build())).isEqualTo(true);
    assertThat(ISTEXT.SELF.apply(null, Grid.builder().add(0, 0, 22000.000).build())).isEqualTo(false);
    assertThat(ISTEXT.SELF.apply(null, Grid.builder().add(0, 0, true).build())).isEqualTo(false);
    assertThat(ISTEXT.SELF.apply(null,
        Grid.builder()
            .add(0, 0, "Yes")
            .add(0, 1, false)
            .build()))
        .isEqualTo(true);
    assertThat(ISTEXT.SELF.apply(null, Grid.builder().add(0, 0, new NAException()).build())).isEqualTo(false);
  }

  @Test
  public void test_errorFromArgumentsMismatch() {
    assertThat(ISTEXT.SELF.apply(null)).isEqualTo(new NAException());
    assertThat(ISTEXT.SELF.apply(null, "A", "B")).isEqualTo(new NAException());
  }
}

File: src/test/js/formulas/info/ISTEXTTest.ts
import { assert } from "chai";
import { it, describe } from "../../testutils/TestUtils";
import { stub } from "sinon";
import { DivException } from "../../../../main/js/errors/DivException";
import { F7Exception } from "../../../../main/js/errors/F7Exception";
import { F7ExceptionName } from "../../../../main/js/errors/F7ExceptionName";
import { NAException } from "../../../../main/js/errors/NAException";
import { RefException } from "../../../../main/js/errors/RefException";
import { ISTEXT } from "../../../../main/js/formulas/info/ISTEXT";
import { Grid } from "../../../../main/js/models/common/Grid";
import { CollateralLookupFunction, LookupFunction } from "../../../../main/js/models/common/Types";
import { CommonModels } from "../../testutils/CommonModels";

describe("ISTEXT", function () {
  it("should work with numbers", function () {
    assert.deepEqual(ISTEXT.SELF.run(null, 10), false);
  });

  it("should work with strings", function () {
    assert.deepEqual(ISTEXT.SELF.run(null, "String"), true);
  });

  it("should work with booleans", function () {
    assert.deepEqual(ISTEXT.SELF.run(null, true), false);
    assert.deepEqual(ISTEXT.SELF.run(null, false), false);
  });

  it("should work with errors", function () {
    assert.deepEqual(ISTEXT.SELF.run(null, new NAException()), false);
    assert.deepEqual(ISTEXT.SELF.run(null, new DivException()), false);
    assert.deepEqual(ISTEXT.SELF.run(null, new RefException()), false);
  });

  it("should work with blank/null/empty", function () {
    assert.deepEqual(ISTEXT.SELF.run(null, null), false);
  });

  it("should work with lookup", function () {
    const lookup = stub();
    const collateralLookup = stub();
    const F = new ISTEXT(lookup as LookupFunction, collateralLookup as CollateralLookupFunction);
    collateralLookup.withArgs(CommonModels.A1, CommonModels.M22_RANGE).returns(10);
    assert.deepEqual(F.run(CommonModels.A1, CommonModels.M22_RANGE), false);
    assert.isTrue(collateralLookup.calledWith(CommonModels.A1, CommonModels.M22_RANGE));
    assert.equal(collateralLookup.callCount, 1);
    assert.isTrue(lookup.notCalled);
  });

  it("should work with grid", function () {
    assert.deepEqual(ISTEXT.SELF.run(null, Grid.builder().add(0, 0, "Nope").build()), true);
    assert.deepEqual(ISTEXT.SELF.run(null, Grid.builder().add(0, 0, 22000).build()), false);
    assert.deepEqual(ISTEXT.SELF.run(null, Grid.builder().add(0, 0, true).build()), false);
    assert.deepEqual(
      ISTEXT.SELF.run(null, Grid.builder().add(0, 0, new DivException()).build()),
      false
    );
  });

  it("should return error when arguments are not of the correct length", function () {
    assert.deepEqual((ISTEXT.SELF.run(null) as F7Exception).name, F7ExceptionName.NA);
    assert.deepEqual((ISTEXT.SELF.run(null, "A", "B") as F7Exception).name, F7ExceptionName.NA);
  });
});
